airflow/jobs/scheduler_job_runner.py

```diff
@@ -1760,18 +1760,19 @@ def _verify_integrity_if_dag_changed(self, dag_run: DagRun, session: Session) ->
 
         Return True if we determine that DAG still exists.
         """
-        latest_version = SerializedDagModel.get_latest_version_hash(dag_run.dag_id, session=session)
-        if dag_run.dag_hash == latest_version:
+        latest_dag_version = DagVersion.get_latest_version(dag_run.dag_id, session=session)
+        latest_dag_version_id = latest_dag_version.id if latest_dag_version else None
+        if dag_run.dag_version_id == latest_dag_version_id:
```

Comment:
If I understand the logic correctly, `dag_run.dag_version_id` can naturally be None. What would this None value mean? Something like “version -1” i.e. the state when Airflow has DAG versioning? And does this None state interact with this check?

Maybe all these questions can be summed into a more fundamental thought, can all DAGs just always have a version (and `latest_dag_version` is never None here) instead of having a special “no version” state?

Reply:
`dag_run.dag_version_id` cannot be none. This check for whether `latest_dag_version.id` is None is not really necessary, but it was done because of type checking. 

>Maybe all these questions can be summed into a more fundamental thought, can all DAGs just always have a version (and latest_dag_version is never None here) instead of having a special “no version” state?

Yes, all dags must have a version. The typing for `get_latest_version` was made to return dag_version or none for public usage, where someone can give it a non-existing DAG_ID, but I can convert it to a private method if we prefer that. WDYT?

Reply:
I see. Maybe an alternative approach would be to not return None when the function is passed a non-existing DAG, but raise an exception instead? I am not sure if that’s a better design.

Without changing the function though, it’s probably easier to understand code here if it’s written like so

```python
latest_dag_version = DagVersion.get_latest_version(dag_run.dag_id, session=session)
if latest_dag_version is None:
    raise RuntimeError(f"DAG {dag_run.dag_id} does not exist; this should not happen")
...
```

We can also use `if TYPE_CHECKING` and `assert` instead to minimise runtime impact.

Reply:
I decided to use `if TYPE_CHECKING` since we cannot have a non-existing dag_version inside the scheduler. A PR is coming that will remove the deletion of SerializedDAG with `airflow dags reserialize` command.